Add trait `FromStrHex` for tuple structs with single `u32` member

connormullett opened the pull request:
Closes: #1112 

 - Adds new trait `FromStrHex` with 2 methods: `from_hex_str` and `from_hex_str_no_prefix`
 -  Impl new trait on each tuple struct with single u32 member. eg `Time(u32)`
 
As stated in the issue, grep through codebase with `\(u32\)` and `\(pub u32\)` to see all implementations and verify none were missed. 

NonStandardSighashType is an error type and should never be constructed from a hex string. Therefore, it has been omitted from this change.

Tests are somewhat redundant, but cover 4 cases each. 2 happy paths, 1 for each function. 1 case for malformed/invalid hex input, and 1 for calling no_prefix without a prefix

apoelstra commented:
As a high level observation, this adds a new module to `util/` which we're trying to clear out. But I"m no sure where a better location for it might be.

Eventually we will move it to the bitcoin-str crate.

connormullett commented:
Yea, I wasn't exactly sure where to stick it so I picked the most obvious. Would it make sense for it to be in `bitcoin/src/parse.rs` for now?

looks like its failing the no-std tests, fixing those up now

connormullett commented:
Force push: rebase to resolve conflicts and squash

apoelstra reviewed (approved):
ACK 8e369d52419f07da125a4bc75d06b28a93483f82

Kixunil commented on bitcoin/src/blockdata/locktime/absolute.rs:
```diff
@@ -86,6 +87,15 @@ impl fmt::Display for PackedLockTime {
 }
 
+impl FromHexStr for PackedLockTime {
+    fn from_hex_str_no_prefix<S: AsRef<str> + Into<String>>(s: S) -> Result<Self, FromHexError>
```
From my experience types that impl `AsRef<str>` often also impl `Into<String>`. Even if not, people can probably use `&*`.

So in practice it shouldn't cause noticeable degradation of ergonomics while being more performant.

apoelstra replied:
Ok, sure, that makes sense.

Kixunil commented on bitcoin/src/blockdata/locktime/absolute.rs:
```diff
@@ -86,6 +87,15 @@ impl fmt::Display for PackedLockTime {
+    fn from_hex_str_no_prefix<S: AsRef<str>>(s: S) -> Result<Self, FromHexError>
+        where Self: Sized
+    {
+        let sequence = u32::from_str_radix(s.as_ref(), 16).map_err(|e| FromHexError::ParseHex(e.to_string()))?;
```
FYI `from_str_radix` may be slow but I didn't measure. Totally OK for now and someone can file a PR later if better perf is needed.

connormullett replied:
For sure, this seemed to be the easiest and most readable way to do this that I found without adding more dependencies

Kixunil commented on bitcoin/src/blockdata/locktime/absolute.rs:
```diff
@@ -750,4 +779,124 @@ mod tests {
+    fn time_from_str_hex_happy_path() {
+        let hex = "0xba70d";
+        let time = Time::from_hex_str(hex).unwrap();
+        assert_eq!(time, Time(763_661));
```
Nit: would be easier to review if this was hex literal too. :)

connormullett replied:
resolved for all tests, just using a const now for each set of tests

Kixunil commented on bitcoin/src/util/from_hex_str.rs:
```diff
@@ -0,0 +1,44 @@
+
+/// Trait that allows types to be initialized from hex strings
+pub trait FromHexStr {
+    /// Parses provided string as hex requiring 0x prefix.
```
Missing `type Error;` which I believe we should have in case some integer newtypes have invariants.

connormullett replied:
added

Kixunil commented on bitcoin/src/util/from_hex_str.rs:
```diff
@@ -0,0 +1,44 @@
+
+/// Hex Parsing error
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[non_exhaustive]
```
I suspect we may not need `non_exhaustive` here but let's keep it for now.

connormullett replied:
:+1:

Kixunil commented on bitcoin/src/util/from_hex_str.rs:
```diff
@@ -0,0 +1,44 @@
+#[non_exhaustive]
+pub enum FromHexError {
+    /// Parsing Error
+    ParseHex(String),
```
I noticed you're converting inner error to a string. Please don't do that - that loses a bunch of information. I suggest one of these options:

```rust
pub enum FromHexError<E> {
    /// The input was not a valid hex string.
    ParseHex(E),

    /// The input is missing `0x` prefix.
    ///
    /// The field `0` contains the input that was attempted to be parsed.
    MissingPrefix(String),
}
```

```rust
pub enum FromHexError<E> {
    /// The input is not a valid hex string.
    ///
    /// The field `0` contains the input that was attempted to be parsed.
    // The field `1` contains the inner error.
    ParseHex(String, E),
    /// The input is missing `0x` prefix.
    ///
    /// The field `0` contains the input that was attempted to be parsed.
    MissingPrefix(String),
}
```

The former requires a newtype for each error but may be better when users use `no_prefix` method because then they don't have to deal with the  `MissingPrefix` branch. IOW the former is nicer to consumers but will be a bit annoying for us unless we come up with some more ideas to improve it.

connormullett replied:
I've figured we don't need the `MissingPrefix` error variant when we can just check if it starts with the prefix and pass `s` as is to the no prefix method. The only way for the no prefix method to fail is the call to from_str_radix which raises the question if we need the custom error type at all (which I still think we do)

connormullett replied:
basically, any valid hex representation (with prefix or not) will work and only needs the implementation for conversion of the non-prefix path. Which I think makes it easier

Kixunil reviewed (changes requested):
I find error handling insufficient (see comments) and also adding new things to `util` while we're removing it is not helpful. `parse` is a good place for this.

Kixunil commented on bitcoin/src/parse.rs:
```diff
@@ -120,3 +120,40 @@ macro_rules! impl_parse_str_through_int {
+            s.as_ref()
+        } else {
+            s.as_ref().trim_start_matches("0x")
+        };
```
This is not done according to the design we discussed in the associated issue. If you ignored it on purpose, please provide your reasoning. Also the doc comment doesn't match.

connormullett replied:
Yea this was kind of an epiphany when changes things up based on your suggestion. The reasoning is the conversion from a hex string shouldn't really care about the prefix, but the implementer should care about how the conversion is performed. This gets rid of one of the error variants so I thought it made sense

connormullett replied:
I'll update the doc comment in the following set of changes

Kixunil replied:
> conversion from a hex string shouldn't really care about the prefix

In the associated issue we considered it footgun to make the prefix optional because e.g `10` is not the same thing in dec and hex but is a valid number for both. In this library we're focused on providing APIs that lower the risk of mistakes.

> the implementer should care about how the conversion is performed

For things that are always hex we already have `FromStr`. `FromHexStr` is for cases where it needs to do something else than `FromStr`.

connormullett replied:
I agree with what you're saying but in the issue, providing no 0x prefix creates an error. This error was never constructed in any of the implementations I provided so the linter was going crazy. Then I thought, if we are already allowing the user to provide the implementation for NO prefix, we can just trim it or pass it to that function. This limits the mistakes on the users end but in theory means that someone implementing this or using this functionality may mix-up prefix/no-prefix. I went with the former as it was the happiest path with less errors and got rid of an entire variant

connormullett replied:
Basically, the way its implemented now, any hex string whether prefix or not works with the least amount of errors (which I think is good). And the no_prefix path is there as a way to be implementation specific with no prefix. That way, its very clear what the inputs are to the function. Any hex string, and a hex string without a prefix

apoelstra replied:
Thanks for that link @Kixunil! I was just thinking about this the other day but couldn't remember what the principle was called.

And +1 to the general idea that in this library, we definitely prioritize footgun elimination over "it just works TM".

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -1389,6 +1401,29 @@ mod tests {
+    #[test]
+    fn compact_target_from_hex_str_happy_path() {
+        let actual = CompactTarget::from_hex_str(HEX_LITERAL).unwrap();
+        let expected = CompactTarget(u32::from_str_radix(HEX_LITERAL.trim_start_matches("0x"), 16).unwrap());
```
This is not what I meant. I meant to let rustc parse the integer by writing `CompactTarget(0x01003456)`

connormullett replied:
Oh duh, lol

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -269,6 +271,16 @@ impl From<CompactTarget> for Target {
 }
 
+impl FromHexStr for CompactTarget {
+    type Error = ParseIntError;
```
This is not carrying the input making error messages hard to understand. It should be OK to use `crate::parse::ParseIntError` instead.

connormullett replied:
:+1:

connormullett commented:
@Kixunil Latest commit will show the lint warnings I was talking about. The error is only ever used in the trait implementation and not in any other implementation and throws an error because of the lint level. 

The error from my end
![image](https://user-images.githubusercontent.com/41083414/203661130-978cb3ed-87dd-4110-b045-6bba544ed2a8.png)

connormullett commented:
Looks like the MissingPrefix error is because the module isn't public in lib.rs. It looks like this is intended (no module documentation, and mostly all internal helpers). Could this module be made public to let the compiler know that its ok?

Kixunil commented:
Oh, yes, we need both the error type and the trait public. I'm not sure about publishing it under `parse`. It's not a bad name but maybe there's better? @apoelstra @tcharding any ideas?

Kixunil commented:
I did consider `str` but I'm a bit concerned some tools could confuse it for the type.

apoelstra commented:
Yeah, that's reasonable. We could call it `bitcoin_str` I guess. I'm not super attached to any particular name.

connormullett commented:
Moved to a new file. Added license header and some basic documentation. Documentation is not my strong suit so suggestions are appreciated as I'm sure y'all would have more insight into what this module will be like in the future

Kixunil commented:
`bitcoin::bitcoin_str` sounds weird. I'd rather use parse if we don't have anything better.

apoelstra commented:
I don't like `parse` because it sounds too general (may cover consensus encodings etc) and too specific (is only about parsing, not serialization, not conversion between formats, etc). I want to emphasize that these are human-readable encodings.

`bitcoin::bitcoin_str` does sound weird but "we wanted to avoid conflicts with std's two uses of `str`" is a pretty good reason for weirdness.

Kixunil commented:
`FromHexStr` is only about parsing not serialization, so the name makes sense. :)

I'd personally rather take the risk and call it `str` than `bitcoin_str` if `parse` is not acceptable.

apoelstra commented:
> `FromHexStr` is only about parsing not serialization, so the name makes sense. :)

True, but the module we're proposing will eventually become the `bitcoin-str` crate, and contain more things!

> I'd personally rather take the risk and call it str than bitcoin_str if parse is not acceptable.

Ok. I don't have a strong opinion either way. We could alse use `string` since lowercase `string` is not used anywhere else, or `human_enc` or something.

tcharding commented:
+1 for `string.rs` for the module name.

connormullett commented:
> the PR should generally not include patches that change code written in previous patches

This is confusing me. My general workflow is to commit and push changes through the PR, each commit after a review is me implementing those suggested changes. Then afterwards, I'll rebase to squash it down. Am I supposed to be doing this another way?

tcharding commented:
> Then afterwards, I'll rebase to squash it down. 

No worries, if that's your workflow I can work around that.

tcharding commented on bitcoin/src/bitcoin_string.rs:
```diff
@@ -0,0 +1,50 @@
+/// Hex parsing error
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[non_exhaustive]
+pub enum FromHexError<E> {
+    /// The input was not a valid hex string.
+    ///
+    /// The field `0` contains the inner error.
+    /// The field `1` contains the input that was attempted to be parsed.
+    ParseHex(E, String),
+    /// The input is missing `0x` prefix.
+    ///
+    /// The field `0` contains the input that was attempted to be parsed.when calling from_hex_str
+    MissingPrefix(String)
+}
```
```suggestion
pub enum FromHexError<E> {
    /// The input was not a valid hex string, contains the error and also the invalid input.
    ParseHex(E, String),
    /// The input is missing `0x` prefix, contains the invalid input.
    MissingPrefix(String)
}
```

connormullett replied:
:+1:

tcharding commented on bitcoin/src/bitcoin_string.rs:
```diff
@@ -0,0 +1,50 @@
+
+/// Trait that allows types to be initialized from hex strings
+pub trait FromHexStr {
+    /// An error occured while parsing from a hex string.
```
```suggestion
    /// An error occurred while parsing the hex string.
```

connormullett replied:
:+1:

tcharding commented on bitcoin/src/blockdata/locktime/absolute.rs:
```diff
@@ -86,6 +87,17 @@ impl fmt::Display for PackedLockTime {
+    type Error = core::num::ParseIntError;
+
+    fn from_hex_str_no_prefix<S: AsRef<str> + Into<String>>(s: S) -> Result<Self, FromHexError<Self::Error>>
+        where Self: Sized
```
```suggestion
    where
        Self: Sized
```

Can you run the formatter once you are ready to merge please. We are adding a new file here and not excluding it from `rustfmt` in `rustfmt.toml` so `cargo +nightly fmt` will format it for you.

connormullett replied:
Sure, I somehow broke my format on save and completely forgot it existed. Will make sure this is done in the future. Thanks

tcharding replied:
Be aware that we use the nightly toolchain to format so if you have your editor configured with the stable toolchain then you'll get loads of unwanted changes on save.

tcharding commented on bitcoin/src/blockdata/locktime/absolute.rs:
```diff
@@ -750,4 +785,64 @@ mod tests {
+    }
+
+    #[test]
+    fn time_from_str_hex_invalid_hex_should_ergr() {
```
```suggestion
    fn time_from_str_hex_invalid_hex_should_err() {
```

connormullett replied:
:+1:

connormullett commented:
> No worries, if that's your workflow I can work around that.

Let me rephrase that because I'd rather make it easier for you guys and learn something new. I didn't understand that comment because of the workflow I use.

> Said in another way, the PR should generally not include patches that change code written in previous patches

What exactly does this mean? How am I supposed to change the code I'm submitting with a PR if I'm not supposed to change it within the PR?

tcharding commented:
Thanks for re-asking, I'll try to explain a bit more clearly. Sorry if I say too much :)

Each patch [0] should make a single change to the codebase. A PR may be multiple patches. If during review a suggestion is made and you wish to implement the change you should rebase and make the change to the original patch that introduced the change. Then you can force push to update the PR. Sometimes review suggests some change that can be done as a separate patch on top but its not super common in my experience, often if this happens a second PR is better.

The benefit of this approach is that any reviewer that comes along can pull the patch set down to their local machine and go through each patch one by one and review them.

Please ask again if I've not explained it perfectly, I'm more than happy spending the time, its time well invested :) 

[0] patch and commit are synonyms

tcharding commented:
In case you are wondering how reviewers review the force pushes, one can pull the latest version then use for example `git range-diff c05188a...@` to see what has changed. If one uses the commit hash of the previously reviewed version then you can see what has changed with the force push.

connormullett commented:
> If one uses the commit hash of the previously reviewed version then you can see what has changed with the force push.

That was exactly what confused me and this cleared it up. I'll be sure to do it this way from here on out. Thanks for the help!

tcharding commented:
This still needs squashing into a single commit bro. And just because its related, we don't want two commits to ever be merged with exactly the same commit logs. Since we are discussing git foo here is a blog post that explains how to write git commit messages, doing so makes reviewing better because
- it forces the dev to understand exactly what they are doing
- it makes it easy for reviewers to see if the dev understands exactly what they are doing (especially when implementing review suggestions)
- it catches mistakes when the dev meant to do one thing but did something else
- it makes it easier for the reviewer to get context on the proposed change in the PR

https://cbea.ms/git-commit/

connormullett commented:
> This still needs squashing into a single commit bro

Sorry, thought I got all of them and looks like I didn't. 

> we don't want two commits to ever be merged with exactly the same commit logs

So when I go to rebase and It has all the squashed commits messages, I should keep them for the description part of the commit? and leave the general message (the first line) as something like what I have currently?

Kixunil commented:
Your commit still says "Fix rebase issue. Squash entire PR down to 1 commit" instead of description of the actual code change. :(

I really like the explanation that git commits should tell a story of how the software was developed. Things like "typo fixes" are not interesting (unless those typos were already in `master`) but things like "Refactor X to prepare Y" deserve their own commit. In your case a single commit makes most sense. However it should tell about the whole change you did, not just the last squash. (The title is still good though.)

connormullett commented:
> Your commit still says "Fix rebase issue. Squash entire PR down to 1 commit" instead of description of the actual code change. :(

I know, there wasn't anything to compare and there wasn't a change. This was to resolve the issue of my mistake previously. Where I missed a commit while rebasing. This won't happen again btw now that I'm aware of the process


> So when I go to rebase and It has all the squashed commits messages, I should keep them for the description part of the commit? and leave the general message (the first line) as something like what I have currently?

This still hasn't been answered

apoelstra commented:
>So when I go to rebase and It has all the squashed commits messages, I should keep them for the description part of the commit? and leave the general message (the first line) as something like what I have currently?

Usually only one commit will have the "real" commit message, so you can delete everything but that. In other cases, *none* of them are an accurate message for the combined commit, so you need to rewrite it from scratch.

tcharding commented:
> This won't happen again btw now that I'm aware of the process

No stress bro, we are all learning no matter how much we know. If I had a dollar for every mistake I made I'd be fiat rich right now :) Everyone here is more than happy to help you out, then your next contribution will be even more valuable.

tcharding reviewed (dismissed):
ACK 9403a1bded6bc8aac587e6cfc2d035c6e709366f

tcharding commented:
Well done @connormullett, this PR is hot now! 

If you want to learn about macros, a future improvement could be macro-izing the implementations of `FromHexStr` (for ones that use similar logic). One can loose a lot of time playing with macros but the more you do it the faster you get and they are super useful.

connormullett commented on bitcoin/src/blockdata/locktime/absolute.rs:
```diff
@@ -86,6 +87,17 @@ impl fmt::Display for PackedLockTime {
 }
 
+impl FromHexStr for PackedLockTime {
+    type Error = core::num::ParseIntError;
```
I'm having trouble implementing this. Can't figure out how to convert from the std/core `ParseIntError` to `parse::ParseIntError`. I've tried implementing From, which became a pain because of the generic, every type of map_err I could think of, and more. I believe I went to implement this at one point but ran into similar issues. I don't know how to resolve this and just opted to use core::num::ParseIntError.

Removed the String from FromHexError::ParseHex

connormullett replied:
Also, crate::parse::ParseIntError holds the core::num::ParseIntError, and from_str_radix returns std::num::ParseIntError

Kixunil replied:
I suggest for starters add `pub(crate) fn hex_u32` in `parse` module that is similar to `parse::int` and call `crate::parse::hex_u32` instead of `u32::from_str_radix` in all those places. We can clean it up (make it generic) later.

Kixunil replied:
for any type `path::Foo` `core::path::Foo` is always the same type as `std::path::Foo` so don't worry about it, just write `core` instead of `std` everywhere.

connormullett replied:
Ok, that makes sense. Thanks for clarifying

tcharding replied:
Hey Connor, you've pushed a new version which implies request for review but you haven't addressed the comment here by @Kixunil about using `crate::parse`. Please either implement his suggestions or comment on why you think this is not a good idea. Definitely do not blindly follow suggestions, anyone is welcome to argue their opinion but we need either a counter-argument or the suggested change to move forward. This is standard procedure in many open source projects. "I don't understand what you are saying here so I left it as is" is also a valid response :)

connormullett replied:
I'm in the process of implementing the change. Once its squashed, it'll be good to go. Pushing is force of habit so I'll hold back the commits as to not request reviews

tcharding replied:
No sweat, legend.

apoelstra commented:
In a23d1ed8fecc27862a1454771e945414c284cb9d:

I'd prefer we call the module `string`. We'll name the eventual crate `bitcoin-string` but it'll be exported from rust-bitcoin just as `string`. (Similar tto what we do with `hashes` now.)

connormullett commented:
> In [a23d1ed](https://github.com/rust-bitcoin/rust-bitcoin/commit/a23d1ed8fecc27862a1454771e945414c284cb9d):
> 
> I'd prefer we call the module `string`. We'll name the eventual crate `bitcoin-string` but it'll be exported from rust-bitcoin just as `string`. (Similar tto what we do with `hashes` now.)

Oh, I thought the suggestion was to replace str with string. My mistake I'll fix it

connormullett commented:
Rebase to resolve merge conflicts

connormullett commented:
@tcharding I still can't figure out this git strategy, I'm creating commits that have meaningful messages and after the force push from squashing they just disappear (because they've been squashed). This is what I'm doing

 - creating commits using the style in the link you sent earlier
 - rebasing with `git rebase -i HEAD~<num commits>` and using `squash` in all but the oldest which stays at default `pick`
 - write the new commit message that summarizes the whole PR

I see in another PR you wrote, your single commits aren't going away after force pushing. How can I do this? I can't find anything online nor in the git book that helps solve this issue

tcharding commented:
`squash` squashes _up_ so you have
```
pick
pick 
squash
```
To squash the last commit _up_ into the 2nd commit.

Kixunil commented:
I personally perform squashing by combination of `git reset --soft HEAD~N` and `git commit --amend` where `N` is the number of commits I want to suash (if PR has three commits `N` is 2). Much easier. :)

tcharding commented:
Little known tip, you can use `@` instead of `HEAD`

apoelstra reviewed (approved):
ACK f0eb24751966285b9c933436c8c67750d2c092ef

Kixunil commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,47 @@
+    /// This is intended for user-supplied inputs or already-existing protocols in which 0x prefix is used.
+    fn from_hex_str<S: AsRef<str> + Into<String>>(s: S) -> Result<Self, FromHexError<Self::Error>>
+    where
+        Self: Sized,
```
Oh, there should be `: Sized` bound on the trait, not here. The compiler didn't suggest the right thing in this case. Sorry for only noticing now :(

connormullett replied:
It's ok, better we find it now than later

Kixunil commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,47 @@
+    /// It should be only used for existing protocols which always encode values as hex without 0x prefix.
+    fn from_hex_str_no_prefix<S: AsRef<str> + Into<String>>(s: S) -> Result<Self, Self::Error>
+    where
+        Self: Sized;
```
Same as above.

connormullett replied:
:+1:

Kixunil commented on bitcoin/src/parse.rs:
```diff
@@ -84,6 +84,15 @@ pub(crate) fn int<T: Integer, S: AsRef<str> + Into<String>>(s: S) -> Result<T, P
+pub(crate) fn hex_u32<S: AsRef<str> + Into<String>>(s: S) -> Result<u32, ParseIntError> {
+    u32::from_str_radix(s.as_ref(), 16).map_err(|error| ParseIntError {
+        input: s.into(),
+        bits: u8::try_from(core::mem::size_of::<u32>() * 4).expect("max is 32 bits for u32"),
```
This needs to be `* 8`: number of bits in a byte. (You could just put `32` there but it's nice to have it ready for making it generic.)

connormullett replied:
Good catch

connormullett commented:
> I personally perform squashing by combination of `git reset --soft HEAD~N` and `git commit --amend` where `N` is the number of commits I want to suash (if PR has three commits `N` is 2). Much easier. :)

Ah so instead of squashing into one commit that already existed, you're creating a new commit based on the squash of the commits. This is exactly what I was looking for, thanks so much and sorry about the git confusion. Never worked on a project with this style and makes a lot of sense to why you guys use it. Thanks

sanket1729 commented on bitcoin/src/blockdata/locktime/absolute.rs:
```diff
@@ -482,6 +494,18 @@ impl fmt::Display for Height {
+        where Self: Sized
+    {
+        let height = crate::parse::hex_u32(s)?;
+        Ok(Self(height))
```
Is there any way we prevent this sort of direct constructor initialization? One idea is to move structs in private modules, but that seems ugly

connormullett replied:
Thank you for bringing this up because this gave me a revelation that should make this look a lot better given your suggestion. Appreciate it

Kixunil replied:
>  move structs in private modules, but that seems ugly

It doesn't seem ugly to me.

sanket1729 commented on bitcoin/src/blockdata/locktime/absolute.rs:
```diff
@@ -565,6 +589,17 @@ impl fmt::Display for Time {
+        where Self: Sized
+    {
+        let time = crate::parse::hex_u32(s)?;
+        Ok(Self(time))
```
Same comment here. Use explicit constructor from_consensus

connormullett replied:
:+1:

sanket1729 commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -404,6 +405,17 @@ impl Sequence {
+        where Self: Sized
+    {
+        let sequence = crate::parse::hex_u32(s)?;
+        Ok(Self(sequence))
```
This is okay for sequence. As a nit and for consistency, still good to use the constructor here.

connormullett replied:
:

sanket1729 commented on bitcoin/src/pow.rs:
```diff
@@ -269,6 +271,18 @@ impl From<CompactTarget> for Target {
+        Self: Sized,
+    {
+        let compact_target = crate::parse::hex_u32(s)?;
+        Ok(Self(compact_target))
```
Same comment

sanket1729 commented on bitcoin/src/sighash.rs:
```diff
@@ -471,6 +472,18 @@ impl fmt::Display for NonStandardSighashType {
+        Self: Sized,
+    {
+        let sighash = crate::parse::hex_u32(s)?;
+        Ok(Self(sighash))
```
Same comment here

connormullett replied:
NonStandardSighashType doesn't have an implementation of from_consensus

sanket1729 reviewed (commented):
Thanks for your contribution. Overall this PR looks great.

Found some bugs in the implementation. Any ideas on how we can restrict the default constructor apart from throwing the definition in a module? We have had the same bug in rust-miniscript were we incorrectly used the default constructor instead of new that returned a result

tcharding commented on bitcoin/src/amount.rs:
```diff
@@ -1469,8 +1470,14 @@ mod verification {
         kani::assume(n1.checked_sub(n2).is_some()); // assume we don't overflow in the actual test
-        assert_eq!(SignedAmount::from_sat(n1) + SignedAmount::from_sat(n2), SignedAmount::from_sat(n1 + n2));
-        assert_eq!(SignedAmount::from_sat(n1) - SignedAmount::from_sat(n2), SignedAmount::from_sat(n1 - n2));
+        assert_eq!(
+            SignedAmount::from_sat(n1) + SignedAmount::from_sat(n2),
+            SignedAmount::from_sat(n1 + n2)
+        );
+        assert_eq!(
+            SignedAmount::from_sat(n1) - SignedAmount::from_sat(n2),
+            SignedAmount::from_sat(n1 - n2)
+        );
```
These are formatter changes that are unrelated to this PR. Since you have to rebase to fix CI can you remove these also please.

connormullett replied:
Yeah, I'm trying to figure out how I messed up and broke CI. It looks completely unrelated to what I implemented and not sure how that got in the PR

sanket1729 commented:
@connormullett , this PR is not breaking the CI. It is already broken :( . There is a hotfix #1432 that should fix it.

connormullett commented:
> @connormullett , this PR is not breaking the CI. It is already broken :( . There is a hotfix #1432 that should fix it.

Thanks for the headsup, thought I messed up while catching up to master

Kixunil commented on bitcoin/src/sighash.rs:
```diff
@@ -471,6 +472,15 @@ impl fmt::Display for NonStandardSighashType {
     }
 }
 
+impl FromHexStr for NonStandardSighashType {
```
Oh, this is an error type so this implementation doesn't make sense.

connormullett replied:
Oh didn't realize that. I'll omit the implementation and make note in the PR

connormullett commented:
force push: rebase + remove NonStandardSighashType's implementation of FromHexStr

connormullett commented:
Is CI still broken @tcharding? Looks like some other pull requests are failing as well

Kixunil commented on bitcoin/src/blockdata/locktime/absolute.rs:
```diff
@@ -565,6 +585,16 @@ impl fmt::Display for Time {
+
+    fn from_hex_str_no_prefix<S: AsRef<str> + Into<String>>(s: S) -> Result<Self, Self::Error> {
+        let time = crate::parse::hex_u32(s)?;
+        dbg!(time);
```
The CI isn't broken, you forgot to remove this one.

connormullett commented:
Force push: remove forgotten debug statement

Kixunil reviewed (dismissed):
ACK 62086a2ae303f65b9aad95b74c0caba09c622924

tcharding commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,42 @@
+/// Hex parsing error
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[non_exhaustive]
+pub enum FromHexError<E> {
```
This error type is missing implementations of `Display` and `std::error::Error` (you can see `blockdata::locktime::absolute::Error` for examples of these). While you are at it, if its not too much bother can you put the error below the trait. I'm on a personal mission to put the most important things first in files :)

connormullett replied:
Yea, that makes sense

Kixunil replied:
Damn, I feel quite embarrassed about not catching this one. :D

tcharding replied:
Yeah I thought the same when I saw it, I actually didn't believe that we would have missed it and re-read the whole PR thread :)

tcharding commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,42 @@
+        if !s.as_ref().starts_with("0x") {
+            Err(FromHexError::MissingPrefix(s.into()))
+        } else {
+            Self::from_hex_str_no_prefix(s.as_ref().trim_start_matches("0x"))
+                .map_err(FromHexError::ParseHex)
```
This could be
```rust
            Ok(Self::from_hex_str_no_prefix(s.as_ref().trim_start_matches("0x"))?)
```
If we had
```rust

impl<E> From<E> for FromHexError<E> {
    fn from(e: E) -> Self {
        FromHexError::ParseHex(e)
    }
}
```

connormullett replied:
For sure, this'll clean it up nicely. Added

tcharding reviewed (changes requested):
`From` impl is up to you but I think we should definitely have the `Display` and `std::error::Error`. Thanks for sticking with this, I know we are a picky bunch :)

connormullett commented:
> Thanks for sticking with this, I know we are a picky bunch :)

Don't worry about it, gives me something to do in my free time haha. This PR's been a rough ride but it looks to be nearing done

tcharding reviewed (dismissed):
ACK c13051de433301b14cd8fe85abb488f6ec65be2f

tcharding commented:
Solid! Thanks.

Kixunil commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,77 @@
+
+        match *self {
+            ParseHex(ref e) => write_err!(f, "failed to parse lock time from string"; e),
+            MissingPrefix(ref value) => write_err!(f, "input is missing an `0x` prefix"; value),
```
The inner value is not an error but an input. So it should be `write!(f, "the input value `{}` is missing the `0x` prefix")`

tcharding replied:
ouch, now I'm the goose that missed _this_ one :)

tcharding replied:
I think the lesson here, at least for me, is that a second review is very useful, sometimes I'm lazy and skip reviews if others have done them already  - lesson learned.

connormullett replied:
Done

Kixunil commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,77 @@
+#[cfg_attr(docsrs, doc(cfg(feature = "std")))]
+impl<E> std::error::Error for FromHexError<E>
+where
+    E: fmt::Debug + std::error::Error + 'static,
```
Nit: `std::error::Error` implies `Debug` so it shouldn't be needed.

connormullett replied:
Oh for sure, I didn't realize this was a thing

connormullett commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,77 @@
+
+impl<E> From<E> for FromHexError<E> {
+    fn from(e: E) -> Self { FromHexError::ParseHex(e) }
+}
```
Oops I forgot this in the last push, will include this in a second

Kixunil commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,76 @@
+        use self::FromHexError::*;
+
+        match *self {
+            ParseHex(ref e) => write_err!(f, "failed to parse lock time from string"; e),
```
Crap, only now have I noticed this message says `lock time` but it's supposed to be generic. It should say something like "failed to parse hex string"

connormullett replied:
Oh rip, looks like a copy/paste error. My bad

apoelstra commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,76 @@
+    ///
+    /// This is **not** recommended for user-supplied inputs because of possible confusion with decimals.
+    /// It should be only used for existing protocols which always encode values as hex without 0x prefix.
+    fn from_hex_str_no_prefix<S: AsRef<str> + Into<String>>(s: S) -> Result<Self, Self::Error>;
```
I think this comment is fine. I think "existing protocols which always encode values as hex without 0x prefix" reads as a thinly-veiled jab at the user-readable "asm" output from Bitcoin Core. I wouldn't mind if we were more explicit honestly.

I agree that as written this suggests "don't use non-prefixed hex" which seems like a strong and unjustified recommendation. What we mean to say is "don't be ambiguous, like Core's asm output is" which is pretty reasonable :)

Anyway what's written is fine IMO.

apoelstra commented on bitcoin/src/string.rs:
```diff
@@ -0,0 +1,76 @@
+/// Hex parsing error
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub enum FromHexError<E> {
+    /// The input was not a valid hex string, contains the error that occurred while parsing.
```
maybe "not a valid hex-encoded object" or "not valid hex or not a valid object".

connormullett replied:
Maybe another variant specifically for overflow?

apoelstra commented:
Looks good to me! I think it's a little silly to be defensive against the case that `sizeof(u32)` is not 4 but that's fine.

apoelstra reviewed (approved):
ACK e00dfa980615d57446215eae58e63689f48f0c7d

apoelstra commented:
Ah, I see!
